### PR TITLE
Issue #309: allow external template files

### PR DESCRIPTION
**Pull request by rnveach**
Issue #309

We are allowing external files so each repository can define its own template and not have to wait on this project to update it if a change is needed.

Sevntu's xdoc template will be:
````
  <#escape x as x?html>
    <div id="${releaseNo}" class="spoiler pre_opened_spoiler">
    <div class="name"><span><h3>Release ${releaseNo}</h3></span></div>
    <div class="text">
      <#if breakingMessages?has_content>
      <p>Breaking backward compatibility:</p>
        <ul>
        <#list breakingMessages as message>
          <li>
            ${message.title}.
	    Author: ${message.author}<#if message.issueNo != -1>
	    <a href="https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/${message.issueNo}">#${message.issueNo}</a></#if>
          </li>
        </#list>
        </ul>
      </#if>
      <#if newMessages?has_content>
      <p>New:</p>
        <ul>
          <#list newMessages as message>
          <li>
            ${message.title}.
	    Author: ${message.author}<#if message.issueNo != -1>
	    <a href="https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/${message.issueNo}">#${message.issueNo}</a></#if>
          </li>
          </#list>
        </ul>
      </#if>
      <#if bugMessages?has_content>
      <p>Bug fixes:</p>
        <ul>
          <#list bugMessages as message>
          <li>
            ${message.title}.
	    Author: ${message.author}<#if message.issueNo != -1>
	    <a href="https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/${message.issueNo}">#${message.issueNo}</a></#if>
          </li>
          </#list>
        </ul>
      </#if>
      <#if notesMessages?has_content>
      <p>Notes:</p>
        <ul>
          <#list notesMessages as message>
          <li>
            ${message.title}.
	    Author: ${message.author}<#if message.issueNo != -1>
	    <a href="https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/${message.issueNo}">#${message.issueNo}</a></#if>
          </li>
          </#list>
        </ul>
      </#if>
    </div>
    </div>
  </#escape>
````

Other templates will come later.

### CODE REVIEW

**Comment by rnveach** [comment]
Failure is not related to changes again.

**Comment by romani** [comment]
hmm, smth is going on ....
from appveyor
```
urls[63] = file:/C:/Users/appveyor/.m2/repository/commons-io/commons-io/1.4/commons-io-1.4.jar
Number of foreign imports: 1
import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
-----------------------------------------------------
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:164)
	... 20 more
Caused by: java.lang.NoClassDefFoundError: org/apache/maven/doxia/siterenderer/DocumentContent
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
	at java.lang.Class.getDeclaredConstructors(Class.java:2020)
```

I saw such stuff in main repo in Distelli CI - https://pipelines.puppet.com/checkstyle/builds/418540

File: releasenotes-builder/src/main/java/com/github/checkstyle/Main.java [hunk 1]
```diff
@@ -167,24 +168,69 @@ private static void runPostGeneration(Multimap<String, ReleaseNotesMessage> rele
                 TemplateProcessor.getTemplateVariables(releaseNotes, releaseNumber);
 
         if (cliOptions.isGenerateAll() || cliOptions.isGenerateXdoc()) {
+            final boolean external = cliOptions.getXdocTemplate() != null;
```
Comment (romani) [inline releasenotes-builder/src/main/java/com/github/checkstyle/Main.java]: lets avoid "external" term.
There should be default template and provided by user.
I would get rid  of such temporal variable at all.

We just need to give to TemplateProcessor some stream to template, or just give it Template instance, or ... .
Main class should orchestrate and play with options. TemplateProcessor should just be configured in exact way and do generation.

File: releasenotes-builder/src/main/java/com/github/checkstyle/TemplateProcessor.java [hunk 2]
```diff
@@ -52,20 +54,34 @@ private TemplateProcessor() { }
...
+
+        if (externalTemplate) {
+            final File file = new File(templateFilename);
+
+            fileName = file.getName();
+            configuration.setTemplateLoader(new FileTemplateLoader(file.getParentFile()));
+        }
+        else {
+            fileName = templateFilename;
+            configuration.setClassForTemplateLoading(Main.class, "/" + TEMPLATE_FOLDER_PATH);
```
Comment (romani) [inline releasenotes-builder/src/main/java/com/github/checkstyle/TemplateProcessor.java]: as you move loading of tempalate of out this class, you will resolve cycle dependency of Main and TemplateProcessor.

File: releasenotes-builder/src/main/java/com/github/checkstyle/Main.java (hunk 1 above)
Comment (rnveach) [reply releasenotes-builder/src/main/java/com/github/checkstyle/Main.java]: Done.

File: releasenotes-builder/src/main/java/com/github/checkstyle/TemplateProcessor.java (hunk 2 above)
Comment (rnveach) [reply releasenotes-builder/src/main/java/com/github/checkstyle/TemplateProcessor.java]: Done.

**Comment by rnveach** [comment]
I confirmed locally that files work fine. Tests confirmed resource works fine.

**Comment by rnveach** [comment]
Made switch to string loader.

**Pull request merged**